intermediate_source/forward_ad_usage.py

```diff
@@ -172,5 +172,75 @@ def jvp(ctx, gI):
                          check_backward_ad=False, check_undefined_grad=False,
                          check_batched_grad=False)
 
+######################################################################
+# Functional API
+# --------------------------------------------------------------------
+# We also offer a higher-level functional API in functorch
+# for computing Jacobian-vector products that you may find simpler to use
+# depending on your use case.
+#
+# The benefit of the functional API is that there isn't a need to understand
+# or use the lower-level dual tensor API and that you can compose it with
+# other functorch transforms (like vmap); the downside is that it offers you
+# less control.
+#
+# Note that the remainder of this tutorial will require functorch
+# (https://github.com/pytorch/functorch) to run. Please find installation
+# instructions at the specified link.
+
+import functorch as ft
+
+primal0 = torch.randn(10, 10)
+tangent0 = torch.randn(10, 10)
+primal1 = torch.randn(10, 10)
+tangent1 = torch.randn(10, 10)
+
+def fn(x, y):
+    return x ** 2 + y ** 2
+
+# Here is a basic example to compute the JVP of the above function.
+# The jvp(func, primals, tangents) returns func(*primals) as well as the
+# computed jvp. Each primal must be associated with a tangent of the same shape.
+primal_out, tangent_out = ft.jvp(fn, (primal0, primal1), (tangent0, tangent1))
```

Comment:
I am bit confused on the functorch APIs which ones return functions vs eagerly compute the result.
Is there any doc that explain why both are used depending on the function?

Reply:
- jvp eagerly computes the result
- vjp eagerly computes the output and returns a vjp_fn
- All the other APIs (grad, vmap, jacrev, jacfwd, hessian) return a function
See https://pytorch.org/functorch/functorch.html for more details.

The main rationale for this is to align the API with JAX's so that it is easy to port code. In general the design philosophy is:
- return a function when possible
- for vjp, it doesn't make sense to return a function that computes the output AND the vector-jacobian-product, because if we were to run this function multiple times we would only need to compute the output once
- for jvp: I don't have a good reason why it is like this.

Reply:
Consistency with jax sounds good.
Is it also consistent with hips/autograd?

Reply:
I'm not familiar with HIPS/autograd. Reading their README it looks like they only support one operation, `grad`, and that it is higher order (and consistent with jax.grad and functorch.grad): https://github.com/HIPS/autograd